edit-migrate part2 (#208)
Certora/Documentation: docs/user-guide/patterns/require-invariants.md

> # Require Invariants: Elevating Smart Contract Specification Quality

Reviewer: ```suggestion
# Require Invariants: Proving inter-dependent invariants
```

> The "Require Invariants" design pattern utilizes `requireInvariant` statements strategically to establish 

Reviewer: ```suggestion
The`requireInvariant` statements can be used to establish 
```

crucial conditions that must persist throughout the execution of a smart contract. Let's explore the 

> importance of this design pattern and illustrate its application using the provided example.

Reviewer: ```suggestion
usefulness of the `requireInvariant` statement and illustrate its application using the provided example.
```

    
    // all other states
    assert (clientAmount < depositedAmount) => clientSharesBefore >= clientSharesAfter;
    assert contractSharesBefore <= contractSharesAfter;
}
```

## Importance of Require Invariants:

1. **Ensuring Invariant Consistency:**
   - The `totalSharesLessThanUnderlyingBalance` invariant expands the validation scope to include the 
   underlying balance of the current contract. By utilizing the previously established 
   `totalSharesLessThanDepositedAmount` invariant as a prerequisite, the specification writer ensures that the 
   total shares in the contract remain within the limits of the underlying asset balance. This 
   `requireInvariant` approach effectively eliminates counterexamples in states that have already been proven 
   to be unattainable.

2. **Validation Through Rules:**
    - The `sharesRoundingTripFavoursContract` rule showcases the application of require invariants to verify 
    the behavior of share transactions. Similarly to the invariant example, the `requireInvariant` statements 
    enable the specification writer to disregard counterexamples in states that are not relevant to the 
    intended behavior.

> In conclusion, the "Require Invariants" design pattern, as demonstrated through the provided example, offers a 

Reviewer: What I think would be better, not necessary for now, is to show example run links where we try to remove one of the requireInvariants and get a false violation, and explain it

Author: Added todo

systematic methodology to define, validate, and uphold critical conditions within smart contract 
specifications.

> for more information, please visit the [documentation](https://docs.certora.com/en/latest/docs/cvl/statements.html#requireinvariant)

Reviewer 2: Add a period at the end.

Author: Done

> for more information, please visit the [documentation](https://docs.certora.com/en/latest/docs/cvl/statements.html#requireinvariant).

Reviewer: better to use internal ref (nit)

Author: Done
